erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py

```diff
@@ -335,16 +463,17 @@ def make_gl_entries(self):
 				valuation_tax.setdefault(tax.cost_center, 0)
 				valuation_tax[tax.cost_center] += \
 					(tax.add_deduct_tax == "Add" and 1 or -1) * flt(tax.base_tax_amount_after_discount_amount)
-
 		# item gl entries
 		negative_expense_to_be_booked = 0.0
 		stock_items = self.get_stock_items()
+		warehouse_account = get_warehouse_account()
+		
 		for item in self.get("items"):
 			if flt(item.base_net_amount):
 				account_currency = get_account_currency(item.expense_account)
 				gl_entries.append(
 					self.get_gl_dict({
-						"account": item.expense_account,
+						"account": item.expense_account if not self.update_stock else warehouse_account[item.warehouse]["name"],
```

Comment:
Set warehouse account as expense account if perpetual inventory enabled and update stock checked. Check set_against_expense_account() function.